### DISCUSSION TITLE
Peer Selection Algorithm

### DISCUSSION THREAD

**Discussion by whr819987540** [anacrolix/torrent#874] [General]
Let's assume such a scenario: Peer A could get piece P from multiple peers. Which peer would A select at last? And what strategy is this selection based on? Does this project provide any detailed algorithm? If it doesn't and I would like to provide such peer selection algorithms, where should I put my code in this project?

**Comment by anacrolix** [comment]
The core of the algorithm you can follow from these 2 lines: https://github.com/anacrolix/torrent/blob/87f6cdc1e96fdd846bb9c55cfd5176a17373d4b6/requesting.go#L241-L242.

**Comment by whr819987540** [comment]
This decides the order of requesting pieces which will be scanned later:
https://github.com/anacrolix/torrent/blob/e10ffd133acfb476b1ebd69f9380a4e300ad75b9/requesting.go#L186-L186

For each piece to scan, this appends the indexes of the chunks which belong to the piece to `requestHeap.requestIndexes`.
https://github.com/anacrolix/torrent/blob/e10ffd133acfb476b1ebd69f9380a4e300ad75b9/requesting.go#L215-L215

`next.Requests.lessByValue` decides the order of chunks to request.
https://github.com/anacrolix/torrent/blob/e10ffd133acfb476b1ebd69f9380a4e300ad75b9/requesting.go#L255-L255

My question is even if we have requested a chunk from peer A but haven't received it yet, as the operations above don't modify the result https://github.com/anacrolix/torrent/blob/e10ffd133acfb476b1ebd69f9380a4e300ad75b9/requesting.go#L186-L186, we would request the same chunk again, which is inefficient.

I would like to know how you schedule multiple(maybe four recommended by BitTorrent) peerConns and whether the condition mentioned in my question exists.

**Reply by anacrolix** [reply]
If you keep reading in `applyRequestState`, requests will be cancelled from other peers that have more than their share, and outright stolen from peers if the stealer is downloading faster.

**Comment by luodw** [comment]
@anacrolix hello, there is a case, one seeder node and 10 other nodes p2p download from the seeder at the same time. How 10 nodes random pick piece to download? We use 1.15 version and every node download the piece at the same order, so they can not upload data to each other, and only download from the only seeder. If they download different piece, they can upload data to each other and speed up download.

**Reply by anacrolix** [reply]
That's an interesting case. I think a unit test to show this can't happen would be wise. I'll do a read through of the code to see if that's happening. If so consider it a bug and I'll see about fixing it. It would be a regression because the piece ordering was explicitly randomised in earlier versions.

**Reply by luodw** [reply]
@anacrolix thanks, the piece order is not from one to max, it is really randomised. I mean 10 peers download pieces as the same order(not from 1 to max).

by the way, which tracker do you use? we use opentracker and I think it is too old and more feature can not provide.

**Reply by whr819987540** [reply]
@luodw  chihaya, which is written in Golang as well, might be helpful.

**Reply by anacrolix** [reply]
@luodw I believe you are correct. For distinct clients operating on a torrent with similar piece priorities set, they will request pieces from lowest to highest index within each priority class. It should be randomized by client at the "normal" level and below. Note that if some pieces are rarer, they will be preferred, but for a new torrent, or one that is in a very controlled environment, it might not be sufficient. 
https://github.com/anacrolix/torrent/blob/a4989ec165a6d14049e1f3542fd1ab08bfdfba96/request-strategy/order.go#L35

What do you think?

**Reply by anacrolix** [reply]
@luodw I think I jumped to a conclusion here. Pieces are iterated in order to fit the available storage capacity, however requests are randomized within each priority class, except for readaheads. See https://github.com/anacrolix/torrent/blob/87f6cdc1e96fdd846bb9c55cfd5176a17373d4b6/requesting.go#L158. I think this should show good behaviour.

What version of anacrolix/torrent are you using? The behaviour I just described should be present since v1.43.0. If you are using a version later than that and it doesn't seem correct, let me know!

**Reply by anacrolix** [reply]
@luodw @whr819987540 I have a tracker called aragorn that I might open source. What are your experiences with chihaya, are you happy with it? What features are missing from opentrackr? I use `udp://tracker.bittorrent.cloud:1337/announce`.

**Reply by whr819987540** [reply]
@anacrolix 

First, I think [chihaya](https://github.com/chihaya/chihaya) is out-of-the-box, which just needs some modification in the configuration file. So, I hope your trakcer is easy to use and users could configure it as more as possible.

Second, chihaya supports storing peers' info in memory (redis) instead of disk. I think it's efficient.

Third, as for the missing features, I think some simple visualization of the statistical data could be built in.

**Reply by whr819987540** [reply]
> @luodw I believe you are correct. For distinct clients operating on a torrent with similar piece priorities set, they will request pieces from lowest to highest index within each priority class. It should be randomized by client at the "normal" level and below. Note that if some pieces are rarer, they will be preferred, but for a new torrent, or one that is in a very controlled environment, it might not be sufficient.
> 
> https://github.com/anacrolix/torrent/blob/a4989ec165a6d14049e1f3542fd1ab08bfdfba96/request-strategy/order.go#L35
> 
> What do you think?

@anacrolix  When I am implementing the Rarest First strategy, I have the similar problem. That is: even though the peer A would ask another peer B for B's rarest pieces, the piece possession between peers is similar.

For the pieces with the same rarity level, I also request them from lowest index to highest index, which might be a problem.

So what's your suggestion? Is it recommended to randomize the pieces with the same rarity level?

**Reply by anacrolix** [reply]
@whr819987540 I generate a random ordering per torrent instance and use it for the lifetime of the torrent. This means peers will differ, but not be so random as to waste effort for themselves.

Are you experimenting with anacrolix/torrent or working on another client?

**Reply by luodw** [reply]
@anacrolix Thanks for you reply, We use anacrolix/torrent version 1.15 and I will test v1.43.0 later. I think opentracker lacks peer schedule policy ability. I see opentracker only return peers with limit count, but does not take peer distance and peer download bytes into account.

**Reply by luodw** [reply]
@anacrolix by the way, which version do you suggest for production use?

**Reply by anacrolix** [reply]
@luodw almost always the latest, in this case v1.53.1. I only push tags when I think there is value in people upgrading, usually a significant bug fix or new stable features. If I made a big mistake I will retract the version, this has happened a few times. https://github.com/anacrolix/torrent/blob/e10ffd133acfb476b1ebd69f9380a4e300ad75b9/go.mod#L120-L125

**Reply by anacrolix** [reply]
I thought I mentioned it somewhere else, but all the up to date sticky piece randomization was finalized in v1.43.0. So you need to be using that version or later to get all the benefits I have been talking about.

**Comment by whr819987540** [comment]
> @anacrolix thanks, the piece order is not from one to max, it is really randomised. I mean 10 peers download pieces as the same order(not from 1 to max).
> 
> by the way, which tracker do you use? we use opentracker and I think it is too old and more feature can not provide.

@luodw Hey, where is the randomized piece order? Can you provide where the code is?

**Reply by anacrolix** [reply]
All the links are above.

**Comment by whr819987540** [comment]
> @whr819987540 I generate a random ordering per torrent instance and use it for the lifetime of the torrent. This means peers will differ, but not be so random as to waste effort for themselves.
> 
> Are you experimenting with anacrolix/torrent or working on another client?



> @whr819987540 I generate a random ordering per torrent instance and use it for the lifetime of the torrent. This means peers will differ, but not be so random as to waste effort for themselves.
> 
> Are you experimenting with anacrolix/torrent or working on another client?

@anacrolix 
I modify your project since [this version](https://github.com/whr819987540/torrent/commit/fdb0911e28c68fcb703c00d6b21230894baf1193), probably its release version is 1.42.0.

I use one seeder and three leechers in the LAN. For pieces with the same rarity level, I just order them according to the index. 

Part of the log is:
```bash
[2023-11-21 17:03:07 +0800 DBG] "1_1.pkl": 192.168.124.106:42069 piece bitmap: [215], request indexes: [215] [torrent 41c3e8d64158c7e60c4c4b69497ae8fc13a973ef github.com/anacrolix/torrent requesting.go:286]
[2023-11-21 17:03:07 +0800 DBG] "1_1.pkl": 192.168.124.105:42069 piece bitmap: [8 93 97 215 250 291 292 473], request piece indexes: [291 292 250 93 8 97 473 215] [torrent 41c3e8d64158c7e60c4c4b69497ae8fc13a973ef github.com/anacrolix/torrent requesting.go:286]
```

**Comment by anacrolix** [comment]
I'll close this discussion as it's getting quite scattered. Feel free to open new discussions about any specific point!
